src/exception/exception-service.ts

```diff
@@ -17,104 +17,78 @@
+import _ from 'lodash';
+import { failure, Result, success, ValidationError } from './error-handling';
+import entityExceptionRepository from './repo/entity';
+import programExceptionRepository from './repo/program';
 import {
-  checkCoreField,
-  checkIsValidSchema,
-  checkProgramId,
-  checkRequestedValue,
-  FieldValidators,
-  validateRecords,
-  ValidationResult,
-} from './validation';
-
-const L = loggerFor(__filename);
+  EntityException,
+  EntityExceptionRecord,
+  isArrayOf,
+  isFollowupExceptionRecord,
+  isSpecimenExceptionRecord,
+  OnlyRequired,
+  ProgramException,
+  ProgramExceptionRecord,
+} from './types';
+import { isValidEntityType, normalizeEntityFileType } from './util';
+import { commonValidators, validateRecords } from './validation';
 
 /**
- * records should be validated before using this function
- * explicitly typing from tsv record input to ProgramException
- * @param programId
- * @param records
- * @returns ProgramException
+ * creates exception object with tsv style records
+ * @param param0
+ * @returns valid EntityException array
  */
-const recordsToException = (
-  programId: string,
-  records: ReadonlyArray<ProgramExceptionRecord>,
-): ProgramException => {
-  return {
-    programId,
-    exceptions: records.map(r => ({
-      schema: r.schema,
-      coreField: r.requested_core_field,
-      exceptionValue: r.requested_exception_value as ExceptionValueType,
-    })),
-  };
-};
+const recordsToEntityException = ({
+  programId,
+  records,
+}: {
+  programId: string;
+  records: ReadonlyArray<EntityExceptionRecord>;
+}) => {
+  const exception: OnlyRequired<EntityException, 'programId'> = { programId };
 
-// relates to our TSV cols
-export const programValidators: FieldValidators<ProgramExceptionRecord> = {
-  program_name: checkProgramId,
-  schema: checkIsValidSchema,
-  requested_core_field: checkCoreField,
-  requested_exception_value: checkRequestedValue,
-};
+  if (isArrayOf(records, isSpecimenExceptionRecord)) {
+    exception.specimen = records;
+  } else if (isArrayOf(records, isFollowupExceptionRecord)) {
+    exception.follow_up = records;
+  }
 
-const createResult = ({
-  exception,
-  validationErrors = [],
-  error = { code: '', message: '' },
-  success = false,
-}: Result) => ({
-  exception,
-  error,
-  validationErrors,
-  success,
-});
-
-export type Result = {
-  success?: boolean;
-  error?: { code: string; message: string };
-  exception?: ProgramException | undefined;
-  validationErrors?: ValidationResult[];
+  return exception;
 };
 
-type Service = ({ programId }: { programId: string }) => Promise<Result>;
+/**
+ * normalize before schema validation
+ * tsv record values may contain different casing, validation needs normalized casing
+ * eg. tsv record value may be 'Follow Up', this will not pass schema validation for 'follow_up'
+ * @param records
+ */
+const normalizeRecords = (records: readonly EntityExceptionRecord[]) =>
+  records.map(r => ({
```

Comment:
my I suggest using explicit variable names? not in this function, but in more complicated ones, finding `r` may be difficult. ‘record’ would be both more legible and perfectly descriptive.

Reply:
changing this from "r" to "record"
only place I could find the poor "r" naming, if there's another place you're looking at please let me know

Reply:
I haven't found it anywhere else here, thus far. But having to refactor Arranger, where there's a lot of `x => ..`, I'm all for saving future maintainers preventable headaches.